File: LinkedFists/find-loop.py
#! /usr/bin/python

class Node:

  def __init__(self, value):
    self.value = value
    self.next = None


class LinkedList:

  def __init__(self):
    self.head = None

  def push(self, new_val):
    new_node = Node(new_val)
    new_node.next = self.head
    self.head = new_node

  def print_list(self):
    temp = self.head
    while(temp):
      print(temp.value)
      temp = temp.next

  def find_loop(self):
    slow_p = self.head
    fast_p = self.head
    while(slow_p and fast_p and fast_p.next):
      slow_p = slow_p.next
      fast_p = fast_p.next.next

      if slow_p == fast_p:
        print slow_p.value
        return slow_p

    return None

llist = LinkedList()

llist.push(5)
llist.push(10)
llist.push(18)
llist.push(5)
llist.push(25)
llist.push(15)

llist.print_list()

llist.head.next.next = llist.head
print llist.find_loop().value

File: LinkedFists/find-loop.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node{
  int val;
  struct node * next;
}Node;

Node * findLoop(Node * head);
void printList(Node * head);
void push(Node ** head_ref, int data);
void delete(Node * head);

int main(void){
  Node * head = NULL;
  push(&head, 7);
  push(&head, 3);
  push(&head, 5);
  push(&head, 9);
  push(&head, 10);
  push(&head, 3);

  printList(head);

  head->next->next->next->next = head->next;

  Node * loop_node = findLoop(head);
  printf("Looped node is %d\n", loop_node->val);

  return 0;
}

Node * findLoop(Node * head){
  Node * slow_ptr = head, * fast_ptr = head;

  while(slow_ptr && fast_ptr && fast_ptr->next){
    slow_ptr = slow_ptr->next;
    fast_ptr = fast_ptr->next->next;

    if(slow_ptr == fast_ptr) return slow_ptr;
  }
  return NULL;
}

void push(Node ** head_ref, int data){
    Node * new = (Node *)malloc(sizeof(Node));
    new->val = data;
    new->next = *head_ref;
    *head_ref = new;
}

void printList(Node * head){
  puts("");
  while(head){
    printf("%d->",head->val);
    head = head->next;
  }
  puts("");
}
void delete(Node * head){
  while(head){
    Node * prev = head;
    head = head->next;
    free(prev);
  }
}



